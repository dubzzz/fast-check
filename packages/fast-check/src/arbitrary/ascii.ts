import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { buildCharacterArbitrary } from './_internals/builders/CharacterArbitraryBuilder';
import { indexToPrintableIndexMapper, indexToPrintableIndexUnmapper } from './_internals/mappers/IndexToPrintableIndex';

/**
 * For single ascii characters - char code between 0x00 (included) and 0x7f (included)
 * @remarks Since 0.0.1
 * @public
 */
export function ascii(): Arbitrary<string> {
  return buildCharacterArbitrary(0x00, 0x7f, indexToPrintableIndexMapper, indexToPrintableIndexUnmapper);
}
