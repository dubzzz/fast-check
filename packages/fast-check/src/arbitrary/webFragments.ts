import type { Arbitrary } from '../check/arbitrary/definition/Arbitrary';
import { buildUriQueryOrFragmentArbitrary } from './_internals/builders/UriQueryOrFragmentArbitraryBuilder';
import type { SizeForArbitrary } from './_internals/helpers/MaxLengthFromMinLength';

/**
 * Constraints to be applied on {@link webFragments}
 * @remarks Since 2.22.0
 * @public
 */
export interface WebFragmentsConstraints {
  /**
   * Define how large the generated values should be (at max)
   * @remarks Since 2.22.0
   */
  size?: Exclude<SizeForArbitrary, 'max'>;
}

/**
 * For fragments of an URI (web included)
 *
 * According to {@link https://www.ietf.org/rfc/rfc3986.txt | RFC 3986}
 *
 * eg.: In the url `https://domain/plop?page=1#hello=1&world=2`, `?hello=1&world=2` are query parameters
 *
 * @param constraints - Constraints to apply when building instances (since 2.22.0)
 *
 * @remarks Since 1.14.0
 * @public
 */
export function webFragments(constraints: WebFragmentsConstraints = {}): Arbitrary<string> {
  return buildUriQueryOrFragmentArbitrary(constraints.size);
}
