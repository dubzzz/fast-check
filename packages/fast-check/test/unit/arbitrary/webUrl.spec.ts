import fc from 'fast-check';
import type { WebUrlConstraints } from '../../../src/arbitrary/webUrl';
import { webUrl } from '../../../src/arbitrary/webUrl';
import { URL } from 'url';

import {
  assertProduceCorrectValues,
  assertProduceSameValueGivenSameSeed,
  assertProduceValuesShrinkableWithoutContext,
  assertShrinkProducesSameValueWithoutInitialContext,
} from './__test-helpers__/ArbitraryAssertions';
import { Value } from '../../../src/check/arbitrary/definition/Value';
import { buildShrinkTree, renderTree } from './__test-helpers__/ShrinkTree';
import { relativeSizeArb, sizeArb, sizeRelatedGlobalConfigArb } from './__test-helpers__/SizeHelpers';

import * as WebAuthorityMock from '../../../src/arbitrary/webAuthority';
import * as WebFragmentsMock from '../../../src/arbitrary/webFragments';
import * as WebQueryParametersMock from '../../../src/arbitrary/webQueryParameters';
import * as WebPathMock from '../../../src/arbitrary/webPath';
import { withConfiguredGlobal } from './__test-helpers__/GlobalSettingsHelpers';
import { fakeArbitrary } from './__test-helpers__/ArbitraryHelpers';

function beforeEachHook() {
  jest.resetModules();
  jest.restoreAllMocks();
  fc.configureGlobal({ beforeEach: beforeEachHook });
}
beforeEach(beforeEachHook);

describe('webUrl', () => {
  it('should always use the same size value for all its sub-arbitraries (except webAuthority when using its own)', () => {
    fc.assert(
      fc.property(sizeRelatedGlobalConfigArb, webUrlConstraintsBuilder(), (config, constraints) => {
        // Arrange
        const { instance } = fakeArbitrary();
        const webAuthority = jest.spyOn(WebAuthorityMock, 'webAuthority');
        webAuthority.mockReturnValue(instance);
        const webFragments = jest.spyOn(WebFragmentsMock, 'webFragments');
        webFragments.mockReturnValue(instance);
        const webQueryParameters = jest.spyOn(WebQueryParametersMock, 'webQueryParameters');
        webQueryParameters.mockReturnValue(instance);
        const webPath = jest.spyOn(WebPathMock, 'webPath');
        webPath.mockReturnValue(instance);

        // Act
        withConfiguredGlobal(config, () => webUrl(constraints));

        // Assert
        expect(webPath).toHaveBeenCalledTimes(1); // always used
        expect(webAuthority).toHaveBeenCalledTimes(1); // always used
        const resolvedSizeForPath = webPath.mock.calls[0][0]!.size;
        if (constraints.authoritySettings === undefined || constraints.authoritySettings === undefined) {
          expect(webAuthority.mock.calls[0][0]!.size).toBe(resolvedSizeForPath);
        }
        if (constraints.withFragments) {
          expect(webFragments.mock.calls[0][0]!.size).toBe(resolvedSizeForPath);
        }
        if (constraints.withQueryParameters) {
          expect(webQueryParameters.mock.calls[0][0]!.size).toBe(resolvedSizeForPath);
        }
      }),
    );
  });
});

describe('webUrl (integration)', () => {
  type Extra = WebUrlConstraints;

  const extraParametersBuilder = webUrlConstraintsBuilder;

  const isCorrect = (t: string) => {
    // Valid url given the specs defined by WHATWG URL Standard: https://url.spec.whatwg.org/
    // A TypeError will be thrown if the input is not a valid URL: https://nodejs.org/api/url.html#url_constructor_new_url_input_base
    expect(() => new URL(t)).not.toThrow();
  };

  const webUrlBuilder = (extra: Extra) => webUrl(extra);

  it('should produce the same values given the same seed', () => {
    assertProduceSameValueGivenSameSeed(webUrlBuilder, { extraParameters: extraParametersBuilder() });
  });

  it('should only produce correct values', () => {
    assertProduceCorrectValues(webUrlBuilder, isCorrect, { extraParameters: extraParametersBuilder() });
  });

  it('should produce values seen as shrinkable without any context', () => {
    assertProduceValuesShrinkableWithoutContext(webUrlBuilder, { extraParameters: extraParametersBuilder(true) });
  });

  it('should be able to shrink to the same values without initial context', () => {
    assertShrinkProducesSameValueWithoutInitialContext(webUrlBuilder, {
      extraParameters: extraParametersBuilder(true),
    });
  });

  it.each`
    rawValue
    ${'http://my.domain.org/a/z'}
    ${'http://user:pass@my.domain.org/a/z'}
    ${'http://my.domain.org/a/z?query#fragments'}
  `('should be able to shrink $rawValue', ({ rawValue }) => {
    // Arrange
    const arb = webUrl({
      authoritySettings: { withUserInfo: true },
      withQueryParameters: true,
      withFragments: true,
    });
    const value = new Value(rawValue, undefined);

    // Act
    const renderedTree = renderTree(buildShrinkTree(arb, value, { numItems: 100 })).join('\n');

    // Assert
    expect(arb.canShrinkWithoutContext(rawValue)).toBe(true);
    expect(renderedTree).toMatchSnapshot();
  });
});

// Helpers

function webUrlConstraintsBuilder(onlySmall?: boolean): fc.Arbitrary<WebUrlConstraints> {
  return fc.record(
    {
      validSchemes: fc.constant(['ftp']),
      authoritySettings: fc.record(
        {
          withIPv4: fc.boolean(),
          withIPv6: fc.boolean(),
          withIPv4Extended: fc.boolean(),
          withUserInfo: fc.boolean(),
          withPort: fc.boolean(),
          size: onlySmall ? fc.constantFrom('-1', '=', 'xsmall', 'small') : fc.oneof(sizeArb, relativeSizeArb),
        },
        { requiredKeys: [] },
      ),
      withQueryParameters: fc.boolean(),
      withFragments: fc.boolean(),
      size: onlySmall ? fc.constantFrom('-1', '=', 'xsmall', 'small') : fc.oneof(sizeArb, relativeSizeArb),
    },
    { requiredKeys: [] },
  );
}
