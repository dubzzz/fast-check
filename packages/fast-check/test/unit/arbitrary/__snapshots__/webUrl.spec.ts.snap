// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`webUrl (integration) should be able to shrink http://my.domain.org/a/z 1`] = `
""http://my.domain.org/a/z"
├> "http://domain.org/a/z"
|  ├> "http://aomain.org/a/z"
|  |  ├> "http://a.org/a/z"
|  |  |  ├> "http://a.rg/a/z"
|  |  |  |  ├> "http://a.ag/a/z"
|  |  |  |  |  ├> "http://a.aa/a/z"
|  |  |  |  |  |  ├> "http://a.aa"
|  |  |  |  |  |  ├> "http://a.aa/z"
|  |  |  |  |  |  |  ├> "http://a.aa"
|  |  |  |  |  |  |  ├> "http://a.aa/"
|  |  |  |  |  |  |  |  └> "http://a.aa"
|  |  |  |  |  |  |  ├> "http://a.aa/a"
|  |  |  |  |  |  |  |  ├> "http://a.aa"
|  |  |  |  |  |  |  |  └> "http://a.aa/"
|  |  |  |  |  |  |  |     └> "http://a.aa"
|  |  |  |  |  |  |  ├> "http://a.aa/n"
|  |  |  |  |  |  |  |  ├> "http://a.aa"
|  |  |  |  |  |  |  |  ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  └> "http://a.aa"
|  |  |  |  |  |  |  |  ├> "http://a.aa/h"
|  |  |  |  |  |  |  |  |  ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |  ├> "http://a.aa/e"
|  |  |  |  |  |  |  |  |  |  ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |  |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  ├> "http://a.aa/c"
|  |  |  |  |  |  |  |  |  |  |  ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  |  ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |  |  |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  |  └> "http://a.aa/b"
|  |  |  |  |  |  |  |  |  |  |     ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  |     ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |  |     |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  |     └> "http://a.aa/a"
|  |  |  |  |  |  |  |  |  |  |        ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  |        └> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |  |           └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  └> "http://a.aa/d"
|  |  |  |  |  |  |  |  |  |     ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |     ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |     |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |     └> "http://a.aa/c"
|  |  |  |  |  |  |  |  |  |        ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |        ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |        |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |        ├> "http://a.aa/a"
|  |  |  |  |  |  |  |  |  |        |  ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |        |  └> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |        |     └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |        └> "http://a.aa/b"
|  |  |  |  |  |  |  |  |  |           ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |           ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |           |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |           └> "http://a.aa/a"
|  |  |  |  |  |  |  |  |  |              ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |              └> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |                 └> "http://a.aa"
|  |  |  |  |  |  |  |  |  └> "http://a.aa/g"
|  |  |  |  |  |  |  |  |     ├> "http://a.aa"
|  |  |  |  |  |  |  |  |     ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |     |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |     └> "http://a.aa/f"
|  |  |  |  |  |  |  |  |        ├> "http://a.aa"
|  |  |  |  |  |  |  |  |        ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |        |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |        └> "http://a.aa/e"
|  |  |  |  |  |  |  |  |           ├> "http://a.aa"
|  |  |  |  |  |  |  |  |           ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |           |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |           ├> "http://a.aa/a"
|  |  |  |  |  |  |  |  |           |  ├> "http://a.aa"
|  |  |  |  |  |  |  |  |           |  └> "http://a.aa/"
|  |  |  |  |  |  |  |  |           |     └> "http://a.aa"
|  |  |  |  |  |  |  |  |           ├> "http://a.aa/c"
|  |  |  |  |  |  |  |  |           |  ├> "http://a.aa"
|  |  |  |  |  |  |  |  |           |  ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |           |  |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |           |  └> "http://a.aa/b"
|  |  |  |  |  |  |  |  |           |     ├> "http://a.aa"
|  |  |  |  |  |  |  |  |           |     ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |           |     |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |           |     └> "http://a.aa/a"
|  |  |  |  |  |  |  |  |           |        ├> "http://a.aa"
|  |  |  |  |  |  |  |  |           |        └> "http://a.aa/"
|  |  |  |  |  |  |  |  |           |           └> "http://a.aa"
|  |  |  |  |  |  |  |  |           └> "http://a.aa/d"
|  |  |  |  |  |  |  |  |              ├> "http://a.aa"
|  |  |  |  |  |  |  |  |              ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |              |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |              └> "http://a.aa/c"
|  |  |  |  |  |  |  |  |                 ├> "http://a.aa"
|  |  |  |  |  |  |  |  |                 ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |                 |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |                 ├> "http://a.aa/a"
|  |  |  |  |  |  |  |  |                 |  ├> "http://a.aa"
|  |  |  |  |  |  |  |  |                 |  └> "http://a.aa/"
|  |  |  |  |  |  |  |  |                 |     └> "http://a.aa"
|  |  |  |  |  |  |  |  |                 └> …
|  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  └> …
|  |  |  |  |  |  └> …
|  |  |  |  |  └> …
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;

exports[`webUrl (integration) should be able to shrink http://my.domain.org/a/z?query#fragments 1`] = `
""http://my.domain.org/a/z?query#fragments"
├> "http://domain.org/a/z?query#fragments"
|  ├> "http://aomain.org/a/z?query#fragments"
|  |  ├> "http://a.org/a/z?query#fragments"
|  |  |  ├> "http://a.rg/a/z?query#fragments"
|  |  |  |  ├> "http://a.ag/a/z?query#fragments"
|  |  |  |  |  ├> "http://a.aa/a/z?query#fragments"
|  |  |  |  |  |  ├> "http://a.aa?query#fragments"
|  |  |  |  |  |  |  ├> "http://a.aa#fragments"
|  |  |  |  |  |  |  |  ├> "http://a.aa"
|  |  |  |  |  |  |  |  ├> "http://a.aa#"
|  |  |  |  |  |  |  |  ├> "http://a.aa#ments"
|  |  |  |  |  |  |  |  |  ├> "http://a.aa#nts"
|  |  |  |  |  |  |  |  |  |  ├> "http://a.aa#ts"
|  |  |  |  |  |  |  |  |  |  |  ├> "http://a.aa#s"
|  |  |  |  |  |  |  |  |  |  |  |  ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  ├> "http://a.aa#a"
|  |  |  |  |  |  |  |  |  |  |  |  |  └> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  ├> "http://a.aa#j"
|  |  |  |  |  |  |  |  |  |  |  |  |  ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  ├> "http://a.aa#f"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> "http://a.aa#d"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |  └> "http://a.aa#c"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |     ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |     └> "http://a.aa#b"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |        ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |        └> "http://a.aa#a"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  |           └> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  └> "http://a.aa#e"
|  |  |  |  |  |  |  |  |  |  |  |  |  |     ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  |     └> "http://a.aa#d"
|  |  |  |  |  |  |  |  |  |  |  |  |  |        ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  |        ├> "http://a.aa#a"
|  |  |  |  |  |  |  |  |  |  |  |  |  |        |  └> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  |        └> "http://a.aa#c"
|  |  |  |  |  |  |  |  |  |  |  |  |  |           ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  |           └> "http://a.aa#b"
|  |  |  |  |  |  |  |  |  |  |  |  |  |              ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  |              └> "http://a.aa#a"
|  |  |  |  |  |  |  |  |  |  |  |  |  |                 └> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  ├> "http://a.aa#h"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  |  └> "http://a.aa#g"
|  |  |  |  |  |  |  |  |  |  |  |  |  |     ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  |     └> "http://a.aa#f"
|  |  |  |  |  |  |  |  |  |  |  |  |  |        ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  |        ├> "http://a.aa#a"
|  |  |  |  |  |  |  |  |  |  |  |  |  |        |  └> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  |        ├> "http://a.aa#d"
|  |  |  |  |  |  |  |  |  |  |  |  |  |        |  ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  |        |  └> "http://a.aa#c"
|  |  |  |  |  |  |  |  |  |  |  |  |  |        |     ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  |        |     └> "http://a.aa#b"
|  |  |  |  |  |  |  |  |  |  |  |  |  |        |        ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  |        |        └> "http://a.aa#a"
|  |  |  |  |  |  |  |  |  |  |  |  |  |        |           └> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  |        └> "http://a.aa#e"
|  |  |  |  |  |  |  |  |  |  |  |  |  |           ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  |           └> "http://a.aa#d"
|  |  |  |  |  |  |  |  |  |  |  |  |  |              ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  |              ├> "http://a.aa#a"
|  |  |  |  |  |  |  |  |  |  |  |  |  |              |  └> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  |              └> "http://a.aa#c"
|  |  |  |  |  |  |  |  |  |  |  |  |  |                 ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  |                 └> "http://a.aa#b"
|  |  |  |  |  |  |  |  |  |  |  |  |  |                    ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  |                    └> "http://a.aa#a"
|  |  |  |  |  |  |  |  |  |  |  |  |  |                       └> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |  └> "http://a.aa#i"
|  |  |  |  |  |  |  |  |  |  |  |  |     ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |     └> "http://a.aa#h"
|  |  |  |  |  |  |  |  |  |  |  |  |        ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |        ├> "http://a.aa#a"
|  |  |  |  |  |  |  |  |  |  |  |  |        |  └> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |        ├> "http://a.aa#e"
|  |  |  |  |  |  |  |  |  |  |  |  |        |  ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |        |  ├> "http://a.aa#c"
|  |  |  |  |  |  |  |  |  |  |  |  |        |  |  ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |        |  |  └> "http://a.aa#b"
|  |  |  |  |  |  |  |  |  |  |  |  |        |  |     ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |        |  |     └> "http://a.aa#a"
|  |  |  |  |  |  |  |  |  |  |  |  |        |  |        └> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |        |  └> "http://a.aa#d"
|  |  |  |  |  |  |  |  |  |  |  |  |        |     ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |        |     └> "http://a.aa#c"
|  |  |  |  |  |  |  |  |  |  |  |  |        |        ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |        |        ├> "http://a.aa#a"
|  |  |  |  |  |  |  |  |  |  |  |  |        |        |  └> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |        |        └> "http://a.aa#b"
|  |  |  |  |  |  |  |  |  |  |  |  |        |           ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |        |           └> "http://a.aa#a"
|  |  |  |  |  |  |  |  |  |  |  |  |        |              └> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |        └> "http://a.aa#g"
|  |  |  |  |  |  |  |  |  |  |  |  |           ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |           └> "http://a.aa#f"
|  |  |  |  |  |  |  |  |  |  |  |  |              ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |              └> "http://a.aa#e"
|  |  |  |  |  |  |  |  |  |  |  |  |                 ├> "http://a.aa#"
|  |  |  |  |  |  |  |  |  |  |  |  |                 └> …
|  |  |  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  └> …
|  |  |  |  |  |  └> …
|  |  |  |  |  └> …
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;

exports[`webUrl (integration) should be able to shrink http://user:pass@my.domain.org/a/z 1`] = `
""http://user:pass@my.domain.org/a/z"
├> "http://my.domain.org/a/z"
|  ├> "http://domain.org/a/z"
|  |  ├> "http://aomain.org/a/z"
|  |  |  ├> "http://a.org/a/z"
|  |  |  |  ├> "http://a.rg/a/z"
|  |  |  |  |  ├> "http://a.ag/a/z"
|  |  |  |  |  |  ├> "http://a.aa/a/z"
|  |  |  |  |  |  |  ├> "http://a.aa"
|  |  |  |  |  |  |  ├> "http://a.aa/z"
|  |  |  |  |  |  |  |  ├> "http://a.aa"
|  |  |  |  |  |  |  |  ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  └> "http://a.aa"
|  |  |  |  |  |  |  |  ├> "http://a.aa/a"
|  |  |  |  |  |  |  |  |  ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  └> "http://a.aa/"
|  |  |  |  |  |  |  |  |     └> "http://a.aa"
|  |  |  |  |  |  |  |  ├> "http://a.aa/n"
|  |  |  |  |  |  |  |  |  ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |  ├> "http://a.aa/h"
|  |  |  |  |  |  |  |  |  |  ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |  |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  ├> "http://a.aa/e"
|  |  |  |  |  |  |  |  |  |  |  ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  |  ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |  |  |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  |  ├> "http://a.aa/c"
|  |  |  |  |  |  |  |  |  |  |  |  ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  |  |  ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |  |  |  |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  |  |  └> "http://a.aa/b"
|  |  |  |  |  |  |  |  |  |  |  |     ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  |  |     ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |  |  |     |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  |  |     └> "http://a.aa/a"
|  |  |  |  |  |  |  |  |  |  |  |        ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  |  |        └> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |  |  |           └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  |  └> "http://a.aa/d"
|  |  |  |  |  |  |  |  |  |  |     ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  |     ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |  |     |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  |     └> "http://a.aa/c"
|  |  |  |  |  |  |  |  |  |  |        ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  |        ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |  |        |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  |        ├> "http://a.aa/a"
|  |  |  |  |  |  |  |  |  |  |        |  ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  |        |  └> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |  |        |     └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  |        └> "http://a.aa/b"
|  |  |  |  |  |  |  |  |  |  |           ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  |           ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |  |           |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  |           └> "http://a.aa/a"
|  |  |  |  |  |  |  |  |  |  |              ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  |              └> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |  |                 └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |  └> "http://a.aa/g"
|  |  |  |  |  |  |  |  |  |     ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |     ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |     |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |     └> "http://a.aa/f"
|  |  |  |  |  |  |  |  |  |        ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |        ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |        |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |        └> "http://a.aa/e"
|  |  |  |  |  |  |  |  |  |           ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |           ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |           |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |           ├> "http://a.aa/a"
|  |  |  |  |  |  |  |  |  |           |  ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |           |  └> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |           |     └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |           ├> "http://a.aa/c"
|  |  |  |  |  |  |  |  |  |           |  ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |           |  ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |           |  |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |           |  └> "http://a.aa/b"
|  |  |  |  |  |  |  |  |  |           |     ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |           |     ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |           |     |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |           |     └> "http://a.aa/a"
|  |  |  |  |  |  |  |  |  |           |        ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |           |        └> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |           |           └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |           └> "http://a.aa/d"
|  |  |  |  |  |  |  |  |  |              ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |              ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |              |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |              └> "http://a.aa/c"
|  |  |  |  |  |  |  |  |  |                 ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |                 ├> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |                 |  └> "http://a.aa"
|  |  |  |  |  |  |  |  |  |                 ├> "http://a.aa/a"
|  |  |  |  |  |  |  |  |  |                 |  ├> "http://a.aa"
|  |  |  |  |  |  |  |  |  |                 |  └> "http://a.aa/"
|  |  |  |  |  |  |  |  |  |                 |     └> …
|  |  |  |  |  |  |  |  |  |                 └> …
|  |  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  |  └> …
|  |  |  |  |  |  |  └> …
|  |  |  |  |  |  └> …
|  |  |  |  |  └> …
|  |  |  |  └> …
|  |  |  └> …
|  |  └> …
|  └> …
└> …"
`;
